src/shallow/snapshot/Serialize/Serializer.lua

```diff
@@ -0,0 +1,235 @@
+	else
+		for markerName, marker in pairs(Markers) do
+			if value == marker then
+				return ("Markers.%s"):format(markerName)
+			end
+		end
+
+		error(("Cannot serialize value %q of type %q"):format(
+			tostring(value),
+			valueType
+		))
+	end
+end
+
+function Serializer.getKeyTypeOrder(key)
+	if type(key) == "string" then
+		return 1
+	elseif Type.of(key) == Type.HostEvent then
+		return 2
+	elseif Type.of(key) == Type.HostChangeEvent then
+		return 3
+	elseif key == Ref then
+		return 4
+	else
+		return math.huge
+	end
+end
+
+function Serializer.compareKeys(a, b)
+	-- a and b are of the same type here, because Serializer.sortTableKeys
+	-- will only use this function to compare keys of the same type
+	if Type.of(a) == Type.HostEvent or Type.of(a) == Type.HostChangeEvent then
+		return a.name < b.name
+	else
+		return a < b
+	end
+end
+
+function Serializer.sortTableKeys(a, b)
+	-- first sort by the type of key, to place string props, then Roact.Event
+	-- events, Roact.Change events and the Ref
+	local orderA = Serializer.getKeyTypeOrder(a)
+	local orderB = Serializer.getKeyTypeOrder(b)
+
+	if orderA == orderB then
+		return Serializer.compareKeys(a, b)
+	else
+		return orderA < orderB
+	end
+end
+
+function Serializer.table(tableKey, dict, output)
+	if next(dict) == nil then
+		output:write("%s = {},", tableKey)
+		return
+	end
+
+	output:writeAndPush("%s = {", tableKey)
+
+	local keys = {}
+
+	for key in pairs(dict) do
+		table.insert(keys, key)
+	end
+
+	table.sort(keys, Serializer.sortTableKeys)
+
+	for i=1, #keys do
+		local key = keys[i]
+		local value = dict[key]
+		local serializedKey = Serializer.tableKey(key)
+
+		if type(value) == "table" then
+			Serializer.table(serializedKey, value, output)
+		else
+			output:write("%s = %s,", serializedKey, Serializer.tableValue(value))
+		end
+	end
+
+	output:popAndWrite("},")
+end
+
+function Serializer.props(props, output)
+	Serializer.table("props", props, output)
+end
+
+function Serializer.children(children, output)
+	if #children == 0 then
+		output:write("children = {},")
+		return
+	end
+
+	output:writeAndPush("children = {")
+
+	for i=1, #children do
+		Serializer.snapshotData(children[i], output)
+	end
+
+	output:popAndWrite("},")
+end
+
+function Serializer.snapshotDataContent(snapshotData, output)
+	Serializer.type(snapshotData.type, output)
+	output:write("hostKey = %q,", snapshotData.hostKey)
+	Serializer.props(snapshotData.props, output)
+	Serializer.children(snapshotData.children, output)
+end
+
+function Serializer.snapshotData(snapshotData, output)
+	output:writeAndPush("{")
+	Serializer.snapshotDataContent(snapshotData, output)
+	output:popAndWrite("},")
+end
+
+function Serializer.firstSnapshotData(snapshotData)
+	local output = IndentedOutput.new()
+	output:writeAndPush("return function(dependencies)")
+	output:write("local Roact = dependencies.Roact")
+	output:write("local ElementKind = dependencies.ElementKind")
+	output:write("local Markers = dependencies.Markers")
```

Comment:
Since folks will be committing snapshots, I feel like we should make sure the lua files we generate pass luacheck. We could add the ability to mark dependencies as we see them in the output, though it would require reworking this file a bit.

Reply:
What do we do for the indentation? Right now I indent snapshot files with two spaces to make it more readable, but usually we indent with tabs.  I don't know what we gain from passing luacheck on a generated file too.

Right now there are only warnings showing up

Reply:
That's fair - we can add to the documentation a reminder not to luacheck these. I figured developers tend to luacheck all their checked in lua files, but maybe that is not the case.